### PR TITLE
Add detection of public IP addresses for localhost and existing-server genesis

### PR DESCRIPTION
**Pull request by speedy-beaver**
Challenge: Streisand VPN services not reachable due to external IP not visible on streisand-host for non-identified VM/host/cloud providers
Solution: Compare external and host IP address and update streisand_ipv4_address if neccessary

This will allow installation of streisand VPNs on a larger range of cloud providers. The challenge was discovered when installing streisand on a [Scaleway](https://www.scaleway.com/) 1-XS instances.

### CODE REVIEW

**Comment by speedy-beaver** [comment]
CI build failing seems to be due to external factors. Is it possible to trigger another test run?

**Comment by nopdotcom** [comment]
Sorry about that. If it happens again, comment again to get our attention.

File: playbooks/roles/common/tasks/detect-public-ip.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+- name: "Install dns module"
+  apt:
+    name: dnsutils
+
+- name: "Initialize lookup variable"
+  set_fact:
+    external_ipv4_address: "presumed_failed"
+
+- name: "Check external IP Address through Google"
+  command: dig o-o.myaddr.l.google.com @ns1.google.com TXT +short
```
Comment (nopdotcom) [inline playbooks/roles/common/tasks/detect-public-ip.yml]: Is this behavior documented?

Comment (nopdotcom) [reply playbooks/roles/common/tasks/detect-public-ip.yml]: This also returns an IPv6 address when available.

File: playbooks/roles/common/tasks/detect-public-ip.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
+---
+# detect-public-ip.yml will attempt to identify whether the server's public
+# IP address is different from what is visible on the host and, if
+# successfully detected, ask to update the address for documentation and
+# configuration profiles
+- name: "Install dns module"
```
Comment (nopdotcom) [inline playbooks/roles/common/tasks/detect-public-ip.yml]: Ansible has [the `ipify_facts` module](https://docs.ansible.com/ansible/latest/modules/ipify_facts_module.html), but it has a flaw: it returns IPv6 addresses when available.

Consider using https://v4.ident.me/.json via [the `uri` module](https://docs.ansible.com/ansible/latest/modules/uri_module.html).

Comment (nopdotcom) [reply playbooks/roles/common/tasks/detect-public-ip.yml]: (BTW, this was just the first service I found; I have no special preference for it.)

**Comment by speedy-beaver** [comment]
Thank you for the quick feedback.

## IPv6 address in DNS response
Thank you for noticing. The dig method with the Google servers can be limited to IPv4 by telling dig explicitly to use IPv4 by calling it with ```dig -4```. The pull request has been updated to include this option. When adding IPv6 to streisand, the explicit ```-6``` option will be helpful.

## Google and alternatives, documentation
[Here is a list][1] of lookup-your-own-IP-over-http providers from 2016, over 40% of them fail today. Challenging to pick a stable one based on the names - apart from Akamai maybe.

The best "documentation" for the currently used method available is a person, claiming to work for Google Public DNS, mentioning this method in the [Google Public DNS support forum][2].

Regarding privacy, Google have a comprehensive [privacy statement][3] regarding the collected data.

Given:
a) the limited data provided - essentially we are asking to echo our public IP address back
b) Google promising not to correlate/combine personal information from other services
c) Google running this on top one of the largest distributed global infrastructures
I would propose to stay with this form of resolution and provider until Google stops that service. I am open for feedback.

The CI seems to have problems to reach some download hosts, would be great if you can trigger it again.

[1]: https://unix.stackexchange.com/questions/254328/get-the-external-ip-address-in-shell-without-dig-in-2016
[2]: https://groups.google.com/forum/#!topic/public-dns-discuss/uyzmMcHQBE0 
[3]: https://developers.google.com/speed/public-dns/privacy

**Comment by nopdotcom** [comment]
Sometimes when I resolve this, I get something else.

```
$ dig -4 o-o.myaddr.l.google.com @ns1.google.com TXT +short
"74.12x.xx.xx"
"edns0-client-subnet 192.0.2.72/32"
$
```

Apparently some middlebox is at 74.12x.xx.xx; I'm at "192.0.2.72". There were some reports that the `edns0-client-subnet` field may not be authoritative either. :-(

I think you're right that Akamai is the most likely to be live from that list, but ident.me allows explicit IPv4 queries. How about this—whether you pick https://v4.ident.me/.json or something else that has an explicit IPv4 feature, I'll put a smoke test in CI, and when that service starts blowing up, we'll know it?

**Comment by nopdotcom** [comment]
(I realize all of these alternatives suck. I'm sorry for being picky about this one thing. `dig +short myip.opendns.com @resolver1.opendns.com A` may be everything we're looking for except for the privacy policy, and personally I'm not troubled by that, but I'm not the sole decider.)

**Comment by speedy-beaver** [comment]
No worries, I am very happy that you take this serious and that you are helping. Your additional comments and tests will make the final solution more robust. 

Regarding the privacy policy, my argument is still that we are asking to echo our own IP back, so very little is exposed - I admit there might be aspects even with that which I have not grasped.

I like the DNS based solution, it feels more elegant and slimmer. I am happy to update it to the opendns one for now. 

If you feel one of the http-based lookups is a better fit, I can update that without too much additional efforts (famous last words :) as I have prototyped that solution as well.

**Comment by speedy-beaver** [comment]
Great, thank you for the feedback and help!

**Pull request merged**